File: 30daysOFcode_C++/day 09.cpp
#include <bits/stdc++.h>

using namespace std;

// Complete the factorial function below.
int factorial(int n) {
int fact=1;
while(n>0){
fact=fact*n;
n--;
} return fact;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    int n;
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    int result = factorial(n);

    fout << result << "\n";

    fout.close();

    return 0;
}

File: 30daysOFcode_Go/solution_set2/day 09.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// Complete the factorial function below.
func factorial(n int32) int32 {
	// handling edges cases
	if n == 0 || n == 1 {
		return 1
	}
	// initializing with 1 becuase by default it gets initialized with 0
	var factorial int32 = 1
	// using loop to calculate factorial
	for n > 0 {
		factorial = factorial * n
		n--
	}
	return factorial
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 1024*1024)

	nTemp, err := strconv.ParseInt(readLine(reader), 10, 64)
	checkError(err)
	n := int32(nTemp)

	result := factorial(n)

	fmt.Fprintf(writer, "%d\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
